### PR TITLE
chore(website): [playground] use languageService for linting code

### PR DESCRIPTION
**Pull request by armano2**
## PR Checklist

- [x] Addresses an existing open issue: fixes #000
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

> extracted from #6656 

- migrate custom program creation to languageService
- correct issues with race conditions and double lint
- use tsvfs in broader scope to create program, store and sync files content
- create clear boundaries between editor, linter and parser

### CODE REVIEW

**Comment by armano2** [comment]
test cases:

### Type Checking:

/play/#ts=4.7.1-rc&sourceType=module&code=PTAEFpK6ZhYAUCUBnALgJwJYGM3gCMB7IgGwFMBDAO3HIA8AHDclFLI6lRZWPvxIiwAzUAApMAV3KgAZLNABGUAGolASlABvUAF9EAdwAWWCuOGVSKGQB8boAESUHqxwQeat+hGPoAuUCICACtyPE0AXgA+UClbewsreNB6UAB+WIxpUADE6wBuQSQwflKoHjBqIkJKa3A0avRsagBzCogy0qKHFzUvfNBQZEAZclAqmrqG8CasVqLeTv528eFSIko0WZbwZiIAWyxrbmKOxdgi2oBPahxQYUkbzc47kjFPUBY0SQxqJQHvBiMIgYNB3B54Di-AiUDBvbSIQbCV7qAYjMbVVbrTatHYYfaHVigRDeRAANyIWAAJmNyAZQAAFPEHaxiOHRbR6FFDMCjFZrDZbXH4o5EhCIQHA0H3R6Q0A4FgbciM4XkOFaBEfchfH40unK5mqtkxHS6LnIQCg5JrtdRyNTiUV5VQ0EqmQS3qieejwJiBTjdgbjuKmJKwTLnuhJMJhIo1RqcJx0KBGKAIrqGa6WUaOaaPaBedUHpJrJTwKSYccSQgJSDQFcbqGIeGvlGAEyxhCDSgGShYUE2vUZw2RY2c3OWrs952U+1iqvBmvSxu-IsulViRgBfUEgA8DwA1lUDNQop5dNy817C8XS+Wiiutyz++m11mTeozWBLYxasWikGgTWda3IuTy-JYpDtoMgzxlwoIACSUCmcoKs6D6qh+oDjmwWAtDa04dlBMGJnBBBIY6ipoe655YewuG2kUUFygmoIANqUAANKABAALpIROvbPgaAB04FiCxCGcSR3EYV+P62rERCCQSImkBB6gaoMaLjFeto3hgFZFCI4jkahg5vKe575uAzIriW-oEscximDIYgmauBrmdoZ5adUNnXvZRyIK5KHuW6mgZHEOR3JY1gYVZfm6QFrBBW5lGaHYmTSHFXoJXZg7HGIACEqVme+ubxYctlCgGRSUmEpAwjIIGypQOA4OQjBoCgABqFKUgAwgAQq5BABFm5JUiiiCte1nU9X1Q3BU6oWxeVOWVf5+WIEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMAbXDYciaND7RIAGhWqsq7JCHx4fAO4BlTswDm6MADNTyRHqPHT5iwDlYALY86o4u8G4enlDelv5mQjxIAPI8AFaI3KGuiPrYAL65ALqROAQk5FQ09MiMLKxMfLRO5kL49rSSfAGUbgpQ6praJVBlpMgU1HQMzGwNtN3IsG4AJh2aCyiOkANaurl4RGMTVdN1c7BMS4irAG5C0H3bGrvDB+XjlVM1M-WNPEJuWj4RocKhMBwYJ6DSAqAp5IA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Parsing comments:

/play/#ts=5.0.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLAID2ZArgg1BACZkBuAhoVKgC40EQoVqFUYEMACwAKHn5aCKCgAeYALxgATEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### User defined compiler options

/play/#ts=4.7.1-rc&jsx=true&sourceType=module&showAST=es&code=GYVwdgxgLglg9mABAMTnRAKAlALkQKQGUANAOgFEAbAUwFtqwpEBvAWAChEvEAnaqEDyQAeACYwAbgD5hAenHSOAXw5A&rules=N4XyA&tsConfig=N4XyA

### Should show error on missing libs

/play/#ts=4.6.2&sourceType=module&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwDcctgB5GABRhwFssBnEAClgHMAueJpgSngF4AfIWLA+AHy68BwqrQYgAPERKCePTiuABuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkqyARhWANoC64YAviF0A

### Should deduce libs base on target

/play/#ts=5.0.2&jsx=true&sourceType=module&code=MYewdgzgLgBAHjAvDMBTA7jACgJxAWwEsJUAKASgG4BYAKCA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwo2V4A8pTIB7BogxhQECDHgiAVsmI5JkAHTMcJMQDMyAc3ULEYyDwC+3WdFoBXA4wmYA2j1m4CRUhWp1GqjwAutZQ8Ha0RJLSMpCehCTklDSI9EzMDCJUmQzZEbRUsIj8OMgAJlQiwmKwBcQAFrCMaqjyaJYgFkA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkeAhpgObp4VTokDMk4YAviH5A

### Should parse simple eslint config when pasted

```js
module.exports = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: './tsconfig.json'
  },
  plugins: ['@typescript-eslint'],
  rules: {
    "@typescript-eslint/no-non-null-asserted-optional-chain": "error"
  }
}
```

### Additional checks

- verify if ast is updating 
- verify if page is not crashign when goign from - to playground with navigation
- verify if linting is beeing triggered

**Comment by bradzacher** [comment]
I wonder if we should make it easier to integrate the vfs into our parser by adding a new API?

**Comment by armano2** [comment]
> I wonder if we should make it easier to integrate the vfs into our parser by adding a new API?

that could work, but in our case we actually have to get ast from parser witch currently is done during parsing

https://github.com/typescript-eslint/typescript-eslint/pull/6806/files#diff-d05a371466c38384729239098e16a3a40108ecf9bff740757faca055e308f6edR27-R28

File: packages/website/src/components/lib/createEventHelper.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (JoshuaKGoldberg) [inline packages/website/src/components/lib/createEventHelper.ts:1]: @bradzacher referring back to #6252, what's the recommended approach here? Switching to `...args: unknown[]` causes type errors:

```plaintext
src/components/linter/createLinter.ts:28:37 - error TS2344: Type 'LinterOnParse' does not satisfy the constraint '(...args: unknown[]) => void'.
  Types of parameters 'fileName' and 'args' are incompatible.
    Type 'unknown' is not assignable to type 'string'.

28   const onParse = createEventHelper<LinterOnParse>();
                                       ~~~~~~~~~~~~~
```

Comment (armano2) [reply packages/website/src/components/lib/createEventHelper.ts:1]: in this case we can't use Function or unknown, the only other option that can think of is

```ts
export function createEventHelper<T extends unknown[]>(): {
  trigger: (...args: T) => void;
  register: (cb: (...args: T) => void) => () => void;
} {
  const events = new Set<(...args: T) => void>();

  return {
    trigger(...args: T): void {
      events.forEach(cb => cb(...args));
    },
    register(cb: (...args: T) => void): () => void {
      events.add(cb);
      return (): void => {
        events.delete(cb);
      };
    },
  };
}

```

and use it as 

```ts
  const onLint = createEventHelper<Parameters<LinterOnLint>>();
  const onParse = createEventHelper<Parameters<LinterOnParse>>();
```

but i don't like this :)

**Review by JoshuaKGoldberg — APPROVED** [review]
Ok I'll admit, this is a lot of TS API -intense stuff that I'm not super familiar with. I'm digging into it now and it all seems good... but I'm in a bit of a rush and can't do a full _"read and learn all the things and their context"_ review 😞. So consider this a medium-level-informed LGTM. Assuming I'm not missing any non-obvious gotchas, the code looks great! Awesome! 🔥

File: packages/website/src/components/editor/useSandboxServices.ts [hunk 2]
```diff
@@ -40,52 +41,57 @@ export const useSandboxServices = (
...
           const libs = await worker.getLibFiles();
           for (const [key, value] of Object.entries(libs)) {
-            libEntries.set('/' + key, value);
+            system.writeFile('/' + key, value);
           }
         }
 
-        const system = sandboxInstance.tsvfs.createSystem(libEntries);
+        // @ts-expect-error - we're adding these to the window for debugging purposes
+        window.system = system;
```
[the remark this answers is not in the document]
Comment (armano2) [reply packages/website/src/components/editor/useSandboxServices.ts:83]: we do not want it to be used in code,  if we are keeping it its only for browser debugging

File: packages/website/src/components/editor/LoadedEditor.tsx [hunk 3]
```diff
@@ -21,11 +16,20 @@ import {
...
 
-export interface LoadedEditorProps extends CommonEditorProps {
-  readonly sandboxInstance: SandboxInstance;
-  readonly webLinter: WebLinter;
+function applyEdit(
+  model: Monaco.editor.ITextModel,
+  editor: Monaco.editor.ICodeEditor,
+  edit: Monaco.editor.IIdentifiedSingleEditOperation,
+): void {
+  if (model.isAttachedToEditor()) {
```
[the remark this answers is not in the document]
Comment (armano2) [reply packages/website/src/components/editor/LoadedEditor.tsx:28]: in some palces previosly we where not always computing history to stack, this little helper checks if editor is attached and execute proper function accordingly 

as history i'm talking about ctrl+z/y

File: packages/website/src/components/linter/createLinter.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+  return {
+    rules,
+    triggerFix,
+    triggerLint,
+    updateParserOptions,
+    onParse: onParse.register,
+    onLint: onLint.register,
+  };
+}
```
[the remark this answers is not in the document]
Comment (armano2) [reply packages/website/src/components/linter/createLinter.ts:143]: that what initially we had, there was a class, this is initial refactoring as ideally i want to move it either to service worker or create custom hook

File: packages/website/src/components/linter/createParser.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      compilerHost = createEnv(compilerOptions);
+    },
+    parseForESLint: (
+      text: string,
+      options: ParserOptions = {},
+    ): TSESLint.Linter.ESLintParseResult => {
+      const filePath = options.filePath ?? '/input.ts';
+
+      // if text is empty use empty line to avoid error
+      const code = text || '\n';
```
[the remark this answers is not in the document]
Comment (armano2) [reply packages/website/src/components/linter/createParser.ts:49]: empty string is crashing when used in languageService

File: packages/website/src/components/linter/createLinter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (armano2) [reply packages/website/src/components/linter/createLinter.ts]: i have tried this, and there seem to be issue with how our code is transofmed, we are not transofmring private properties properly and this is not working :(

File: packages/website/src/components/editor/LoadedEditor.tsx [hunk 6]
```diff
@@ -71,287 +76,228 @@ export const LoadedEditor: React.FC<LoadedEditorProps> = ({
...
+        monaco.Uri.parse(webLinter.rules.get(ruleId)?.url ?? ''),
+      );
+      monaco.editor.setModelMarkers(
+        monaco.editor.getModel(monaco.Uri.file(uri))!,
+        'eslint',
+        diagnostics,
+      );
+      updateMarkers();
+    });
+    return () => disposable();
```
[the remark this answers is not in the document]
Comment (armano2) [reply packages/website/src/components/editor/LoadedEditor.tsx:136]: sadly no, this has to stay as is to preserve context of a function

**Pull request merged**